File: 096 Su Doku.pl
#!/usr/bin/perl

# Daniel "Trizen" Șuteu
# License: GPLv3
# Date: 30 January 2017
# https://github.com/trizen

# https://projecteuler.net/problem=96

# Runtime: 13.888s

use 5.016;
use strict;
use integer;

sub check {
    my ($i, $j) = @_;

    my ($id, $im) = ($i / 9, $i % 9);
    my ($jd, $jm) = ($j / 9, $j % 9);

    $jd == $id && return 1;
    $jm == $im && return 1;

    $id / 3 == $jd / 3 and
    $jm / 3 == $im / 3;
}

my @lookup;
foreach my $i (0 .. 80) {
    foreach my $j (0 .. 80) {
        $lookup[$i][$j] = check($i, $j);
    }
}

sub solve_sudoku {
    my ($callback, @grid) = @_;

    sub {
        foreach my $i (0 .. 80) {
            if (!$grid[$i]) {

                my %t;
                undef @t{@grid[grep { $lookup[$i][$_] } 0 .. 80]};

                foreach my $k (1 .. 9) {
                    if (not exists $t{$k}) {
                        $grid[$i] = $k;
                        __SUB__->();
                    }
                }

                $grid[$i] = 0;
                return;
            }
        }

        $callback->(@grid);
        goto SOLUTION_FOUND;
    }->();
    SOLUTION_FOUND: return;
}

open my $fh, '<', 'p096_sudoku.txt'
  or die "Can't open file `p096_sudoku.txt`: $!";
chomp(my @grids = grep { /^[0-9]+$/ } <$fh>);
close $fh;

my $sum = 0;

while (@grids) {
    my @grid = split(//, join('', splice(@grids, 0, 9)));
    solve_sudoku(sub { $sum += "$_[0]$_[1]$_[2]" }, @grid);
}

say $sum;

File: 096 Su Doku.jl
#!/usr/bin/julia

# Daniel "Trizen" Șuteu
# License: GPLv3
# Date: 08 February 2017
# https://github.com/trizen

# https://projecteuler.net/problem=96

# Runtime: 5.685s

function check(i, j)
    id, im = div(i, 9), mod(i, 9)
    jd, jm = div(j, 9), mod(j, 9)

    jd == id && return true
    jm == im && return true

    div(id, 3) == div(jd, 3) &&
    div(jm, 3) == div(im, 3)
end

const lookup = zeros(Bool, 81, 81)

for i in 1:81
    for j in 1:81
        lookup[i,j] = check(i-1, j-1)
    end
end

function solve_sudoku(callback::Function, grid::Array{Int64})
    function solve()
        for i in 1:81
            if grid[i] == 0
                t = Dict{Int64, Bool}()

                for j in 1:81
                    if lookup[i,j]
                        t[grid[j]] = true
                    end
                end

                for k in 1:9
                    if !haskey(t, k)
                        grid[i] = k
                        solve()
                    end
                end

                grid[i] = 0
                return()
            end
        end
        callback(grid)
    end
    solve()
end

function euler_096()

    fh = open("p096_sudoku.txt")
    lines = filter((x)->ismatch(r"^[0-9]+$",x), readlines(fh))

    total = 0

    while length(lines) > 0
        grid = splice!(lines, 1:9)
        grid = map((c)->parse(Int64, c), split(join(map(chomp, grid)), ""))
        solve_sudoku((s)->(total += s[1]*100 + s[2]*10 + s[3]), grid)
    end

    println(total)
end

euler_096()
